drv/mb85rsxx-fram/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+        // > are input to SI. The 3-bit upper address bit is invalid. When 8
+        // > bits of writing data is input, data is written to FRAM memory cell
+        // > array. Risen CS will terminate the WRITE command. However, if you
+        // > continue sending the writing data for 8 bits each before CS rising,
+        // > it is possible to continue writing with automatic address
+        // > increment. When it reaches the most significant address, it rolls
+        // > over to the starting address, and writing cycle keeps on continued
+        // > infinitely.
+
+        // Start the write command.
+        let _cs_is_held_low_while_this_exists =
+            self.0.start_rw_command(Opcode::Write, addr)?;
+        // Wham, bam, write to the FRAM!
+        self.0.spi.write(buf)
+    }
+
+    /// Read bytes from the FRAM starting at `addr` into `buf`.
+    ///
+    /// This is the same as [`Fram::read`], see the documentation for that
+    /// method for details.
+    pub fn read(&self, addr: usize, data: &mut [u8]) -> Result<(), SpiError> {
+        self.0.read(addr, data)
+    }
+
+    /// Reads the FRAM device's product ID.
+    pub fn read_id(&self) -> Result<FramId, SpiError> {
+        self.0.read_id()
+    }
+
+    /// Unset the write enable latch on the FRAM.
+    ///
+    /// If this operation fails, this method returns `Self` so that unsetting
+    /// the write enable latch can be retried.
+    pub fn write_disable(self) -> Result<(), (SpiError, Self)> {
+        match self.0.do_write_disable() {
+            Ok(_) => {
+                // Don't do it again.
+                core::mem::forget(self);
+                Ok(())
+            }
+            Err(e) => Err((e, self)),
+        }
+    }
+}
+
+impl FramId {
+    const MANUFACTURER_FUJITSU: u8 = 0x04;
+
+    fn read<S: SpiServer>(spi: &SpiDevice<S>) -> Result<Self, SpiError> {
+        // Indicates that we must read another two bytes to get the product ID.
+        const CONTINUE: u8 = 0x7f;
+
+        let _cs_is_held_low_while_this_exists = spi
+            .lock_auto(CsState::Asserted)
+            .map_err(|_| SpiError::TaskRestarted)?;
+        let mut buf = [0; 3];
+        spi.exchange(&[Opcode::ReadId as u8, 0, 0], &mut buf)?;
+        let [_, mfg_id, cont] = buf;
+
+        let product_id = if cont == CONTINUE {
+            let mut buf = [0; 2];
+            spi.read(&mut buf)?;
+            u16::from_be_bytes(buf)
+        } else {
+            // no continuation code --- use the bytes we just read.
+            // AFAICT, the Fujitsu FRAM chips don't do this, but other
+            // manufacturers' FRAM chips that use the same protocol do? Would
+            // have to read some more datasheets to be sure.
+            u16::from_be_bytes([mfg_id, cont])
+        };
+
+        Ok(Self { mfg_id, product_id })
+    }
+}
+
+impl<S: SpiServer, const ID: u16> Drop for WritableFram<'_, S, { ID }> {
```

Comment:
Nit: move this to L338, so we're not interleaving `WritableFram` and `FrameId` impl blocks?